instrumentation/apache-log4j-2/src/main/java/com/nr/agent/instrumentation/log4j2/AgentUtil.java

```diff
@@ -59,55 +33,62 @@ public static void recordNewRelicLogEvent(LogEvent event) {
             Throwable throwable = event.getThrown();
 
             if (shouldCreateLogEvent(message, throwable)) {
-                HashMap<String, Object> logEventMap = new HashMap<>(DEFAULT_NUM_OF_LOG_EVENT_ATTRIBUTES);
+                Map<LogAttributeKey, Object> logEventMap = new HashMap<>(DEFAULT_NUM_OF_LOG_EVENT_ATTRIBUTES);
                 if (message != null) {
                     String formattedMessage = message.getFormattedMessage();
                     if (formattedMessage != null && !formattedMessage.isEmpty()) {
-                        logEventMap.put(MESSAGE, formattedMessage);
+                        logEventMap.put(AppLoggingUtils.MESSAGE, formattedMessage);
                     }
                 }
-                logEventMap.put(TIMESTAMP, event.getTimeMillis());
+                logEventMap.put(AppLoggingUtils.TIMESTAMP, event.getTimeMillis());
+
+                if (AppLoggingUtils.isAppLoggingContextDataEnabled() && event.getContextData() != null) {
+                    event.getContextData().forEach((key, value) -> {
```

Comment:
Is this implementation of `forEach` just a simple for loop under the hood?

If not, isn't this much less performant that using a simple loop? For the extra allocation reasons that you pointed out previously? Also based on some experimentation that I've done I found that the execution time is much longer when using streaming operations on small data sets as compared to a simple loop. Unless you expect that there may be many thousands of context attributes I'm not sure of the benefit of streaming.

Reply:
This is a bit of a different case. `event.getContextData()` does not return a Map. It returns a ReadOnlyStringMap which extends Serializable and implements no interfaces.
This ReadOnlyStringMap provides a toMap method and the forEach that can be used here. There are many implementation of that interface, but the one I got in a unit test was SortedArrayStringMap. This impl, will instantiate a new HashMap when toMap is called. The forEach implementation is done using a for(int i...) loop.
So I'd have to do some profiling to see which one allocates less memory.